contrib/python-sdk/openpaisdk/job.py

```diff
@@ -0,0 +1,274 @@
+            "echo \"write config to {}\"".format(c_file),
+            "echo <% $secrets.clusters %> > {}".format(c_file),
+            "opai cluster select <% $secrets.cluster_alias %>",
+        ])
+        # download files
+        cmds += ["opai storage download <% $secrets.work_directory %>/source/{} {}".format(f, f) for f in sources]
+        self.new_deployment(name, pre_commands=cmds, taskRoles=taskRoles)
+
+    def one_liner(self, commands: Union[list, str], image: str, cluster: dict, resources: dict=None, submit: bool=True, **kwargs):
+        self.protocol["prerequisites"].append(Job.new_unit("docker_image", "dockerimage", uri=image))
+        self.add_taskrole("main", {
+            "dockerImage": "docker_image",
+            "resourcePerInstance": resources if resources else self.default_resouces(),
+            "commands": commands if isinstance(commands, list) else [commands]
+        })
+        if submit:
+            return self.submit(cluster["cluster_alias"])
+        else:
+            return self.validate().get_config()
+
+    def from_notebook(self, nb_file: str, image: str, cluster: dict, resources: dict=None, submit: bool=True, interactive_mode: bool=True, token: str="abcd", **kwargs):
+        if not nb_file:
+            interactive_mode, nb_file = True, ""
+        html_file = os.path.splitext(nb_file)[0] + ".html" if not interactive_mode else ""
+        if interactive_mode:
+            cmds = [
+                "jupyter notebook --no-browser --ip 0.0.0.0 --port 8888 --NotebookApp.token={} --allow-root {}".format(token, nb_file),
+            ]
+        else:
+            cmds = [
+                'jupyter nbconvert --ExecutePreprocessor.timeout=-1 --ExecutePreprocessor.allow_errors=True --to html --execute %s' % nb_file,
+                'opai storage upload {} <% $secrets.work_directory %>/output/{}'.format(html_file, html_file),
+            ]
+
+        job_info = self.one_liner(
+            commands = cmds,
+            image=image,
+            resources = resources,
+            cluster = cluster,
+            submit = submit,
+            **kwargs
+        )
+        if not submit:
+            return job_info
+        # post processing after Submitting
+        client = ClusterList().load().get_client(cluster["cluster_alias"])
+        print("waiting job to be started")
+        if interactive_mode:
+            # wait and get ip
+            state = client.wait([self.name], exit_states=["SUCCEEDED", "FAILED", "RUNNING"])[0]
+            assert state == "RUNNING", "why not running {}".format(state)
+            while True:
+                try:
+                    status = client.jobs(self.name)
+                    ip = status["taskRoles"]["main"]["taskStatuses"][0]["containerIp"]
+                except:
+                    ip = None
+                    time.sleep(10)
+                if ip:
+                    break
+            browser_open("http://%s:8888/?token=%s" % (ip, token))
```

Comment:
I think the port 8888 should be added to job config's port list.

Reply:
update it to the port list
```python
resources.setdefault("ports", {})["jupyter"] = 1
cmds = [
    "jupyter notebook --no-browser --ip 0.0.0.0 --port $PAI_CONTAINER_HOST_jupyter_PORT_LIST  --NotebookApp.token={} --allow-root --NotebookApp.file_to_run={}".format(token, nb_file),
]
```
and 
```python
browser_open("http://%s:%s/notebooks/%s" % (status["containerIp"], status["containerPorts"]["jupyter"], nb_file))
```